### PR TITLE
first integration tests

### PR DESCRIPTION
**Pull request by johannbarbie**
Resolves https://github.com/leapdao/leapdao-integ-tests/issues/7

### CODE REVIEW

**Comment by troggy** [comment]
getting error on start

```
troggy@graybox  ~/projects/acebusters/leapdao-integ-tests   feature/integrationTests ●  yarn test
yarn run v1.12.3
$ ./scripts/run.sh
Out folder: 2019-01-17|17:14:14
Starting ganache network...
Deploying contracts...
Starting first node...
Launching next node...
Starting setup...
(node:66891) UnhandledPromiseRejectionWarning: Error: Couldn't decode address from ABI: 0x
```

File: tests/actions/machineGun.js [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+      1000 + Math.round(100 * Math.random()),
+      0,
+      alicePriv
+    );
+    await node.sendTx(transfer1.hex());
+    log('Transfer:', transfer1.hex());
+    log(transfer1.hash());
+    const txData = await node.web3.eth.getTransaction(transfer1.hash());
+    if (!txData) {
+      continue;
```
Comment (troggy) [inline tests/actions/machineGun.js]: what's the idea here?

File: tests/actions/mintAndDeposit.js [hunk 2]
```diff
@@ -0,0 +1,28 @@
+const should = require('chai').should();
+
+module.exports = async function(alice, amount, minter, token, exitHandler, noLog = false) {
+  let log;
+  if (noLog) {
```
Comment (troggy) [inline tests/actions/mintAndDeposit.js]: remove duplication

File: tests/actions/mintAndDeposit.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+    log = function(){};
+  } else {
+    log = console.log;    
+  }
+  
+  log(`------Will mint and deposit ${amount} tokens for ${alice}------`);
+  log(`Minting token to account ${alice}...`);
+  //await token.methods.mint(alice, amount).send({from: minter});
```
Comment (troggy) [inline tests/actions/mintAndDeposit.js]: remove

File: tests/2_depositConsolidateExit.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+    console.log(`${alice} balance after deposit: ${await nodes[0].web3.eth.getBalance(alice)}`);
+    await mintAndDeposit(alice, amount, minter, contracts.token, contracts.exitHandler); //second deposit so there will be 2 utxo to consolidate
+    await sleep(5000);
+    console.log(`${alice} balance after deposit: ${await nodes[0].web3.eth.getBalance(alice)}`);
+    await consolidate(nodes[0], alice);
+    console.log(`${alice} balance after consolidate: ${await nodes[0].web3.eth.getBalance(alice)}`);
+    console.log("Make some more deposits to make sure the block is submitted (with log is off)...");
+    await mintAndDeposit(yyy, 1000, minter, contracts.token, contracts.exitHandler, true);
+    await sleep(sleepTime);
+    for (let i = 0; i < 32; i++) {
```
Comment (troggy) [inline tests/2_depositConsolidateExit.js]: remove duplication across the test cases

File: tests/1_depositTransferExit.js [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+            alicePriv, 
+            bob, 
+            txAmount, 
+            nodes[0]);
+        
+        ((await nodes[0].web3.eth.getBalance(alice)) * 1).should.be.equal(balanceAlice - txAmount);
+        ((await nodes[0].web3.eth.getBalance(bob)) * 1).should.be.equal(balanceBob + txAmount);
+    }
+    console.log("Make some more deposits to make sure the block is submitted (with log is off)...")
+    for (let i = 0; i < 20; i++) {
```
Comment (troggy) [inline tests/1_depositTransferExit.js]: remove duplication across the test cases

File: tests/3_depositExitTransfer.js [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    console.log("║1. Deposit to Alice                       ║");
+    console.log("║2. Exit Alice                             ║");
+    console.log("║3. Try to transfer exited utxo            ║");
+    console.log("╚══════════════════════════════════════════╝");
+    await mintAndDeposit(alice, amount, minter, contracts.token, contracts.exitHandler);
+    await sleep(5000);
+    let plasmaBalanceAfter = (await nodes[0].web3.eth.getBalance(alice)) * 1;
+    console.log(`${alice} balance after deposit: ${plasmaBalanceAfter}`);
+    console.log("Make some more deposits to make sure the block is submitted (with log is off)...")
+    for (let i = 0; i < 32; i++) {
```
Comment (troggy) [inline tests/3_depositExitTransfer.js]: remove duplication across the test cases

File: tests/3_depositExitTransfer.js [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+        await mintAndDeposit(zzz, i + 1, minter, contracts.token, contracts.exitHandler, true);
+        await sleep(1000);
+    }
+    await sleep(3000);
+    console.log("------Exit Alice------");
+    const utxo = await exitUnspent(contracts, nodes[0], alice);
+    console.log("------Attemp to transfer exited utxo from Alice to Bob (should fail)------");
+    let plasmaBalanceBefore = (await nodes[0].web3.eth.getBalance(alice)) * 1;
+    await expect(transferUtxo(utxo, bob, alicePriv, nodes[0])).to.eventually.be.rejectedWith("Non zero error code returned: 2");
+    //log: 'TypeError: Cannot read property \'address\' of undefined'
```
Comment (troggy) [inline tests/3_depositExitTransfer.js]: remove?

File: tests/4_epochLengthExit.js [hunk 8]
```diff
@@ -0,0 +1,53 @@
...
+    const minter = accounts[0].addr;
+    const admin = accounts[1].addr;
+    const alice = accounts[7].addr;
+    const zzz = accounts[9].addr;
+    const amount = 10000000;
+
+    console.log("╔══════════════════════════════════════════╗");
+    console.log("║   Test: Exit after epochLength change    ║");
+    console.log("║Steps:                                    ║");
+    console.log("║1. Deposit to Alice                       ║");
```
Comment (troggy) [inline tests/4_epochLengthExit.js]: should also do transfer as per spec #7

File: run.js [hunk 9]
```diff
@@ -71,19 +77,24 @@ async function run() {
...
   await setup(contracts, nodes, accounts, web3);
   // Wait for setup to propagate to all the nodes
   await sleep(10000);
 
   var testPath = require("path").join(__dirname, "tests");
-  fs.readdirSync(testPath).forEach(async function(test) {
+  const tests = fs.readdirSync(testPath).filter((fileName => {
+    return fs.lstatSync("./tests/" + fileName).isFile();
+  }));
+  for (let i=0; i<tests.length; i++) {
```
[the remark this answers is not in the document]
Comment (eshavkun) [reply run.js]: This is original code from Jan, which I didn't change. If you insist I can do it of course :)

File: src/helpers.js [hunk 10]
```diff
@@ -57,4 +67,58 @@ function makeTransfer(
...
+  txs.forEach((tx, i) => {
+    if (tx.blockNumber > youngestInput.blockNumber) {
+      youngestIndex = i;
+      youngestInput = tx;
+    }
+  });
+  return { index: youngestIndex, tx: youngestInput };
+}
+
+function getYoungestInputTx(web3, tx) {
```
[the remark this answers is not in the document]
Comment (eshavkun) [reply src/helpers.js]: When I was writing this the PR with helpers was not merged yet, so copied them from there to my helpers. Do you want me to remove them and use leap-core?

[the remark this answers is not in the document]
Comment (troggy) [reply src/helpers.js]: > Do you want me to remove them and use leap-core?

Yes, that's what I said in my first comment. You see any problem with this?

**Pull request merged**